design/one-pager-component-mutable-and-versioning.md

```diff
@@ -0,0 +1,475 @@
+- If `alwaysControl` field is true, OAM runtime will always create a new workload when component revision changed
+  and leave the trait to control two workloads rolling update.
+- If `alwaysControl` field is false, this means the trait will control only if revision path specified. OAM runtime
+  will check if no `revisionPath` specified, then it will update workload directly, otherwise it will create new workload. 
+ 
+
+So when we use the trait like below:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  components:
+    - componentName: frontend
+      traits:
+        - trait:
+            apiVersion: core.oam.dev/v1alpha2
+            kind: FancyTrait
+            spec:
+              revision: frontend-c8bb659c5 
+```
+
+The FancyTrait will always apply to the specified revision `frontend-c8bb659c5`, and won't work
+for the other workload.
+
+### Always Using the latest revision when using componentName field
+
+When we use a Component with componentName in ApplicationConfiguration, oam runtime will always apply
+all traits to the latest revision of component and make the old one inactive. 
+
+For example, below is an AppConfig containing one component with two traits.
+
+```yaml
+apiVersion: core.oam.dev/v1alpha2
+kind: TraitDefinition
+metadata:
+  name: rollouts.core.oam.dev
+spec:
+  revision:
+    alwaysControl: true
+    revisonPath:
+      - "spec.source.revision"
+  definitionRef:
+    name: rollouts.core.oam.dev
+---
+apiVersion: core.oam.dev/v1alpha2
+kind: ApplicationConfiguration
+metadata:
+  name: example-appconfig
+spec:
+  components:
+    - componentName: frontend
+      traits:
+        - trait:
+            apiVersion: core.oam.dev/v1alpha2
+            kind: ManuelScaler
+            spec:
+              replica: 3
+        - trait:
+            apiVersion: core.oam.dev/v1alpha2
+            kind: Rollout
+            spec:
+              source:
+                revision: frontend-c8bb659c5 
+```
+
+Once the `frontend` component changed, the `ApplicationConfiguration` controller will trigger
+an upgrade for the application.
+
+1. Create or update the workload to the new component revision according to revision field of binding traits.
+   - if creating new workload, workload name will align with revision name.
+   - if updating workload, workload name will align with the component name.
+   - if a trait who can control revision removed, a new workload will still be created with the same name of component. 
+2. Update all traits pointing their target workloadRef to the new workload.
+   - If there are traits who have revision field and specify to old revision, make sure the workload
+   corresponding to the old revision will still work. 
+3. Periodically do garbage collection for all workloads which don't have any trait specified.
+4. Traits who have revision field should take care of the garbage collection things for the old revision.
```

Comment:
who is responsible for GC? Traits or AppConfig

Reply:
It should be Component Controller do the GC?. Like the relationship between Deployment Revision and Deployment Controller.

Reply:
> who is responsible for GC? Traits or AppConfig

When we have revision control trait binding with an App, then Trait is responsible for GC. Otherwise, the AppConfig controller is responsible.

> It should be Component Controller do the GC? Like the relationship between Deployment Revision and Deployment Controller.

Component Controller won't do the GC things.